src/ray/gcs/gcs_server/gcs_actor_scheduler.cc

```diff
@@ -563,5 +623,45 @@ void RayletBasedActorScheduler::HandleWorkerLeaseReply(
   }
 }
 
+void GcsActorScheduler::HandleWorkerLeaseRejectedReply(
+    std::shared_ptr<GcsActor> actor, const rpc::RequestWorkerLeaseReply &reply) {
+  // The request was rejected because of insufficient resources.
+  if (!actor->GetAcquiredResources().IsEmpty()) {
+    // Return the actor's acquired resources, which updates GCS' resource view.
+    ReturnActorAcquiredResources(actor);
+  }
+  if (normal_task_resources_changed_callback_ &&
+      RayConfig::instance().gcs_actor_scheduling_enabled()) {
+    normal_task_resources_changed_callback_(actor->GetNodeID(), reply.resources_data());
+  }
+  actor->UpdateAddress(rpc::Address());
+  Reschedule(actor);
+}
+
+void GcsActorScheduler::OnActorDestruction(std::shared_ptr<GcsActor> actor) {
+  if (!actor->GetAcquiredResources().IsEmpty()) {
+    ReturnActorAcquiredResources(actor);
+    cluster_task_manager_->ScheduleAndDispatchTasks();
+  }
+}
+
+void GcsActorScheduler::ReturnActorAcquiredResources(std::shared_ptr<GcsActor> actor) {
+  auto &cluster_resource_manager =
+      cluster_task_manager_->GetClusterResourceScheduler()->GetClusterResourceManager();
+  cluster_resource_manager.AddNodeAvailableResources(
+      scheduling::NodeID(actor->GetNodeID().Binary()), actor->GetAcquiredResources());
+  actor->SetAcquiredResources(ResourceRequest());
+}
+
+size_t GcsActorScheduler::GetPendingActorsCount() const {
+  return cluster_task_manager_->GetInfeasibleQueueSize() +
+         cluster_task_manager_->GetWaitingQueueSize();
```

Comment:
It is a good catch here. For gcs scheduler, my idea is making actors stay in the `tasks_to_schedule_` until finding an available node (otherwise actors would go to a raylet for scheduling, which is not gcs scheduling anymore). But now in `cluster_task_manager`, we hard-coded the `exclude_local_node` parameter to be `false` (when calling `GetBestSchedulableNode`), and using the value of `exclude_local_node` for the `require_node_available` parameter of hybrid_policy. So I plan to fix this in another PR, because gcs scheduling is not turned on for the time being.